File: generate_cards.R
# General functions
polar_to_cartesian <- function(r,phi){
	x <- r*cos(phi)
	y <- r*sin(phi)
	return(list(x=x,y=y))
}

figure_perimeter <- function(shape, center_x, center_y, width, height){
	if(shape=='ellipse'){
		a <- width/2
		b <- height/2
		theta <- seq(0,2*pi,length.out=10)
		radius <- (a*b)/(sqrt((b*cos(theta))^2+(a*sin(theta))^2))
		at_origin <- polar_to_cartesian(radius,theta)
		x <- at_origin$x+center_x 
		y <- at_origin$y+center_y
	}
	else if(shape=='diamond'){
		x <- c(center_x-width/2,center_x,center_x+width/2,center_x,center_x-width/2)
		y <- c(center_y,center_y-height/2,center_y,center_y+height/2,center_y)
	}
	else if(shape=='rectangle'){
		x <- c(center_x-width/2,rep(center_x+width/2,2),center_x-width/2,center_x-width/2)
		y <- c(rep(center_y+c(-1,1)*height/2,each=2),center_y-height/2)
	}
	return(list(x=x,y=y))
}

figure_list <- function(n_figures,...){
	if(n_figures==1){
		y_centers <- c(0)
	}
	if(n_figures==2){
		y_centers <- c(-.5,.5)
	}
	if(n_figures==3){
		y_centers <- c(-.5,0,.5)
	}
	fl <- vector(mode='list',length=n_figures)	
	for(f in 1:n_figures){
		fl[[f]] <- figure_perimeter(...,
				center_x=0,center_y=y_centers[f],
				width=1,height=0.25)
	}
	return(fl)
}

draw_perimeters <- function(fl){
	for(f in 1:length(fl)){
		lines(fl[[f]]$x,fl[[f]]$y,
		type='o',pch=21,bg='#ffffff',
		lwd=2,col='#e95698cc')
	}
}

blank_polygons <- function(perimenters){

}


# Figure testing:
try(dev.off())
x11(width=2.5,height=3.5)
par(mar=rep(1,4),xaxs='i',yaxs='i',tck=-0.01,mgp=c(3,0.25,0),
	cex.axis=0.5)
plot(NULL,xlim=c(-1,1),ylim=c(-1,1))
draw_perimeters(figure_list(3,shape='ellipse'))
#fp <- figure_perimeter('rectangle',center_x=0.25,center_y=-0.5,width=1.5,height=1)
#polygon(fp$x,fp$y,border='red',lwd=2,col='#44558889')	


make_card <- function(){
	par(mar=rep(1,4),xaxs='i',yaxs='i')
	plot(NULL,xlim=c(-1,1),ylim=c(-1,1))
	x <- c(-.5,.5,.5,-.5,-.5,-.25,-.25,.25,.25,-.25)
	y <- c(-.5,-.5,.5,.5,-.5,-.25,.25,.25,-.25,-.25)
	polygon(x,y,col='#44589e78',lwd=2,border='#d37834')

	# 0. Choose color ['#color1','#color2','#color3']
	# 1. Fill card with texture ['blank','solid','lines']
	# 2. Draw figure(s) ['ellipse','diamond','rectangle']
	# 	2a. Get figure(s) perimeter coordinates
	# 	2b. Draw blank polygons surrounding figurei(s)
	# 	2c. Draw figure(s) lines

}
png(file='Cards/card.png',width=2.5,height=3.5,units='in',res=500)
make_card()
dev.off()

File: generate_cards.py
import matplotlib.pyplot as plt
import numpy as np

card_height = 3.5
card_width = 2.5
colors = ['#8500ff','#fd7300','#00f6a2']
shapes = ['R','D','E']
numbers = [1,2,3]
shades = ['S','B','T']

def poly_rectangle(center,width=1.25,height=.375):
	poly_points = np.array([
		[center[0]-width/2,center[1]-height/2],
		[center[0]+width/2,center[1]-height/2],
		[center[0]+width/2,center[1]+height/2],
		[center[0]-width/2,center[1]+height/2],
		[center[0]-width/2,center[1]-height/2]]) 
	return poly_points

def poly_diamond(center,width=1.45,height=.425):
	poly_points = np.array([
		[center[0],center[1]-height/2],
		[center[0]+width/2,center[1]],
		[center[0],center[1]+height/2],
		[center[0]-width/2,center[1]],
		[center[0],center[1]-height/2]]) 
	return poly_points

def poly_ellipse(center,a=1.65,b=.15):
	phi = np.linspace(0,np.pi*2,100)	
	r = a*b/np.sqrt(b*np.cos(phi)**2+a*np.sin(phi)**2)
	x = r*np.cos(phi) + center[0]
	y = r*np.sin(phi) + center[1]
	poly_points = np.column_stack((x,y))
	return poly_points


def make_card(col_indx,num,shape,shade):
	# Color
	col = colors[col_indx]
	# Shape
	if shape=='E':	
		poly_fun = poly_ellipse
	elif shape=='R':	
		poly_fun = poly_rectangle
	elif shape=='D':	
		poly_fun = poly_diamond
	# Number
	if num==1:
		y = [0]
	elif num==2:
		y = [-.375,.375]
	elif num==3:
		y = [-.75,0,.75]
	# Fill
	if shade=='B':
		shd = '00'
	elif shade=='S':
		shd = 'ff'
	elif shade=='T':
		shd = '33'
	file_name = shape+'_'+str(num)+'_'+shade+'_'+['p','o','g'][col_indx]+'.svg'
	plt.figure(figsize=(card_width,card_height))
	plt.grid(True)
	for i in range(len(y)):
		polygon = poly_fun(center=[0,y[i]])
		plt.plot(polygon[:,0],polygon[:,1],color=col,lw=3)
		plt.fill(polygon[:,0],polygon[:,1],color=col+shd)
		plt.subplots_adjust(left=0,bottom=0,right=1,top=1)
	plt.xlim(-card_width/2,card_width/2)
	plt.ylim(-card_height/2,card_height/2)
	plt.axis('off')
	plt.savefig('Cards/'+file_name,format='svg',bbox_inches='tight',pad_inches=0)

card_counter = 0
for sp in shapes:
	for sd in shades:
		for cl in range(len(colors)):
			for nm in numbers:
				card_counter = card_counter + 1
				print('Drawing card '+str(card_counter),'\n')
				make_card(col_indx=cl,num=nm,shape=sp,shade=sd)
